theano/tensor/nnet/corr3d.py

```
@@ -215,72 +256,62 @@ def c_code_helper(self, bottom, weights, top, direction, sub,
                 raise ValueError("width must be given for backprop with horizontal sampling or border_mode='half'")
             width = '-1'
         if depth:
             depth = '(*(npy_int64 *)(PyArray_DATA(%s)))' % depth
         else:
-            if ((direction != 0) and (dD != 1)) or ((direction == 1) and (padD == -1)):
+            if ((self.direction != 0) and (self.dD != 1)) or ((self.direction == 1) and (self.padD == -1)):
                 raise ValueError("depth must be given for backprop with depth sampling or border_mode='half'")
             depth = '-1'
-        sub = sub.copy()
-        sub.update(locals())
 
         return """
     // Mandatory args
-    int direction = %(direction)s;  // forward, bprop weights, bprop inputs
+    int direction = %(params)s->direction;  // forward, bprop weights, bprop inputs
 
     // Optional args
-    int dH = %(dH)s;
-    int dW = %(dW)s;
-    int dD = %(dD)s;
-    int dilH = %(dilH)s;
-    int dilW = %(dilW)s;
-    int dilD = %(dilD)s;
-    int padH = %(padH)s;
-    int padW = %(padW)s;
-    int padD = %(padD)s;
+    int dH = %(params)s->dH;
+    int dW = %(params)s->dW;
+    int dD = %(params)s->dD;
+    int dilH = %(params)s->dilH;
+    int dilW = %(params)s->dilW;
+    int dilD = %(params)s->dilD;
+    int padH = %(params)s->padH;
+    int padW = %(params)s->padW;
+    int padD = %(params)s->padD;
 
     PyArrayObject * bottom = %(bottom)s;
     PyArrayObject * weights = %(weights)s;
     PyArrayObject * top = %(top)s;
     PyArrayObject * out2 = NULL;
+    PyArrayObject **out = NULL;
+
+    switch(%(params)s->direction) {
+        case DIRECTION_FORWARD:
+            out = &%(top)s;
+            break;
+        case DIRECTION_BACKPROP_WEIGHTS:
+            out = &%(weights)s;
+            break;
+        case DIRECTION_BACKPROP_INPUTS:
+            out = &%(bottom)s;
+            break;
+        default: break;
```

Default should raise an error I think.